V3 of the challenge (#71)
hashicorp/mktg-webdev-challenge: README.md

[Here's the "people directory" page! 👉](/people)

_**Note:** The above link will only work when opened inside the project running locally._

Use [this Figma file](https://www.figma.com/design/0c70HmKc0rkWCNdP9UYErO) as your source of truth for design and behavior:

_**Note:** If you are unable to inspect elements in the design to get exact values for colors and font styles, make sure you're signed in with a Figma account, as we will expect your submission to match the design as closely as possible._

There are definitely some parts of this that are a little ambiguous, which is how real world projects always work. Treat this the same way you would if you started working with us and this was your first day — this is quite similar to the type of task we would assign a new hire on their first day. We will set up a Discord channel for you to ask questions, get clarification, and collaborate with us in general, and we strongly encourage you to take advantage of this.

> We expect this to be done within 4 hours. We know that you likely have other commitments, and so we generally look at about week for candidates to complete the challenge. If you're approaching 4 hours of total time spent, just stop where you are, and we'll complete the evaluation based on what you have done. The code in your PR(s) should be the same quality as any code you would deploy to production.

Reviewer: `The code in your PR(s) should be the same quality as any code you would deploy to production.` is redundant here. It's in the `"Submitting your work"` section already.

```suggestion
We expect this to be done within 4 hours. We know that you likely have other commitments, and so we generally look at about week for candidates to complete the challenge. If you're approaching 4 hours of total time spent, just stop where you are, and we'll complete the evaluation based on what you have done.
```

The task is to build a "people directory" page. This page should allow visitors to search by name for a person. If this were a real page, we'd want to search all 317 records in our CMS, but for the purposes of this exercise, limit your results to 100. This is the maximum number of results that our CMS API will return in a single page.

> [Here's the "people directory" page! 👉](/people)

Author: This existed in the content.mdx that we removed, so just adding it back

This project uses [Next.js](https://nextjs.org), a react-based framework that generates performant websites that blur the lines between static and dynamic rendering. We have a set of internal tools under the banner of `@hashicorp/platform-*` layered into the config that provides several utilities, and is especially helpful for loading remote data.

Next.js [conventions](https://nextjs.org/learn/basics/fetching-data-for-pages) dictate that an async function called `getStaticProps` can be exported along with any component within the `pages` directory. This function will be executed before the component mounts, passing the returned object into the component as props.

> We will provide you with a `DATO_API_TOKEN` for use in the `.env`, and for Sr. candidates also in the `createDB.ts`, file so that you can load data from a Dato instance. Also the cda-explorer.datocms.com tool we reference below will also need you to put in your `DATO_API_TOKEN` for it to work.

Reviewer: ```suggestion
We will provide you with a `DATO_API_TOKEN` to add to your `.env` file. For Sr. candidates, you will need to add it in `createDB.ts` in order to load data from a DatoCMS.

The cda-explorer.datocms.com tool referenced below should have `<ADD_DATO_API_TOKEN_HERE>` replaced with the `DATO_API_TOKEN` value for it to work.
```

### Building The Page

Once you have the data, the next step is to build out the page! We expect the markup to be clear and semantic, and the styling to be clear, well-structured, and good-looking (we use [CSS Modules](https://nextjs.org/docs/basic-features/built-in-css-support#adding-component-level-css)). Do your best to visually match the styling in the Figma design. Feel free to build out components as part of the page to reduce repetition and break out reasonable chunks of code. The `components` folder can be a home to them.

> Matching person results should update in real-time while typing. Checking the "Hide people missing a profile image" should filter out results that don't have an image. Use the placeholder image from the Figma file for those without an image.

Reviewer: Just realized candidates will be able to see that this was "checking," which may be too much of a clue. :/

This project uses [Next.js](https://nextjs.org), a react-based framework that generates performant websites that blur the lines between static and dynamic rendering. We have a set of internal tools under the banner of `@hashicorp/platform-*` layered into the config that provides several utilities, and is especially helpful for loading remote data.

Next.js [conventions](https://nextjs.org/learn/basics/fetching-data-for-pages) dictate that an async function called `getStaticProps` can be exported along with any component within the `pages` directory. This function will be executed before the component mounts, passing the returned object into the component as props.

> You'll see a similar template provided in the `pages/people` folder, which works out of the box.

Reviewer: a bit confusing given the rivet query package has been removed
```suggestion
```

Use of AI will not be penalized. If you choose to use AI for this project, please include comments about the relevant code as well as your reasoning in your PR. For example, you could write a brief description in the PR comment and/or leave comments on the code generated or modified by AI.

We'll review your PR and leave comments and questions. You're not expected to write any additional code after review, but we'd love to see you reply to review questions to share your thinking. We will review your final submission within 3-4 business days.

## Acceptance Criteria / TL;DR:

- [ ] The department tree creation function does not work. You will need to troubleshoot and fix it for the filtering to work as expected.

> - [ ] You should build all the components you need for this project on your own, within this project. While you should not use external packages we do welcome discussions on where an external package might be a better fit.

Reviewer: ```suggestion
- [ ] You should build all the components you need for this project on your own, within this project. While you should not use external packages, we do welcome discussions on where an external package might be a better fit.
```

- [ ] A simple [Profile component](components/profile/index.tsx) for the people cards has already been created. You may modify this and any other components as needed.
  - [ ] All images on the page can be retrieved from the CMS via URLs from the graphql query.
- [ ] We expect the page to be styled responsively so it looks great on both mobile and desktop.
- [ ] Matching person results should update in real-time while typing.
- [ ] Clicking the "Hide people missing a profile image" should filter users without an image.

> - [ ] If no results match the search, show "No results found" UI from mock

Reviewer: ```suggestion
- [ ] If no results match the search, show "No results found" UI from the Figma design
```
